### DISCUSSION TITLE
Geany needs spyware 🙄

### DISCUSSION THREAD

**Discussion by xiota** [geany/geany#2918] [Ideas]
(This is a strawman proposal for discussion.)

Without spyware, it's difficult to know how many users there are and what plugins and features they use.  Even if having this information couldn't help developers guide development, it would make them feel good knowing that their work is being put to good use.

Anyone have any thoughts about this? Pros and cons?  Implementation details?  Etc.

**Comment by elextr** [comment]
Please note the number of emoji in discussions that mention adding spyware.  

To be clear, I do not support software having spyware, despite the potential benefits for developers.  And as IANAL I don't know what legal issues spyware may have, even if its possible to turn it off.  And anyway statistics tell you what people use now, not what they might use if it was added or an existing unused feature improved.

And really, Geany isn't guided by statistics, but by what people contribute.  No matter how many statistics there are, if nobody contributes a change it won't happen.  

There are no paid directed developers following a master plan shaped by gathered statistics.  Its all whatever people contribute, either their own ideas, or ones they get from feature request issues, or from issue discussions/arguments, or stolen from other projects, or whatever.  Then the only filter on the contributions is if the (volunteer) committers have the time and interest to review test and merge changes.  

If somebody thinks a pull request is not suitable they may simply say so, or suggest changes, or totally ignore it, although getting _no_ response is rare for pull requests, but more so for feature requests.  As people might have noticed one of my bugbears is people who want to effectively remove existing features by replacing them or simply deleting them.  IMO thats like saying "I don't want it that way so nobody can have it" and a better approach is to try to add the new feature as additional functionality, or provide an option to select it (if it can't coexist with an existing feature) or put it in a plugin so the user can decide to include it or not.

In general one or more contributors will comment if an idea is outside their image of what Geany should do, its in an interesting space, a lightweight IDE between plain editors like Vim and full fat IDEs like Visual studio or Eclipse and generally should stay there (IMHO) since there are not many other similar options.  That means there are things it probably will never do (like type inference), and why I use Eclipse for C++ development rather than try to fatten Geany to provide similar features.

Of course as all volunteer time is limited, large changes can sit for a long time before anyone has the time to review, test, and merge them.  Especially in these "interesting" times.  So Geany tends to move in small steps punctuated by occasional lurches, which makes it fun but occasionally frustrating :grin:.

[end thesis, all opinions are my own and other contributors may disagree (and probably will)]

**Reply by xiota** [reply]
> To be clear, I do not support software having spyware...

I've edited the original post to make clear that this is a strawman proposal for discussion.

I also don't like spyware, but maybe an opt-in plugin that reports limited information could be acceptable?

> I don't know what legal issues spyware may have, even if its possible to turn it off.

Many programs "check for updates".  Software from larger companies do more, but of course, they likely have more monetary resources for legal battles than Geany does.

>  And anyway statistics tell you what people use now, not what they might use if it was added or an existing unused feature improved.

Agree.

> And really, Geany isn't guided by statistics, but by what people contribute. No matter how many statistics there are, if nobody contributes a change it won't happen.

Agree.  This is true of smaller open-source programs with limited funds.

> There are no paid directed developers following a master plan shaped by gathered statistics. Its all whatever people contribute, either their own ideas, or ones they get from feature request issues, or from issue discussions/arguments, or stolen from other projects, or whatever. Then the only filter on the contributions is if the (volunteer) committers have the time and interest to review test and merge changes.

Perhaps usage statistics would encourage committers to consider changes that they might not otherwise be motivated to consider?

> If somebody thinks a pull request is not suitable they may simply say so, or suggest changes, or totally ignore it, although getting _no_ response is rare for pull requests, but more so for feature requests. As people might have noticed one of my bugbears is people who want to effectively remove existing features by replacing them or simply deleting them. IMO thats like saying "I don't want it that way so nobody can have it" and a better approach is to try to add the new feature as additional functionality, or provide an option to select it (if it can't coexist with an existing feature) or put it in a plugin so the user can decide to include it or not.

People who ask for a feature to be "changed" would likely be satisfied with an option for their request.

Knowing the usage of a feature would help with determining whether an outright change would be acceptible.  Maybe "no one" is using the feature because it has wonky behavior.

> Of course as all volunteer time is limited, large changes can sit for a long time before anyone has the time to review, test, and merge them. Especially in these "interesting" times. So Geany tends to move in small steps punctuated by occasional lurches, which makes it fun but occasionally frustrating grin.

Maybe usage statistics could encourage more activity on large changes?

**Reply by eht16** [reply]
> > To be clear, I do not support software having spyware...
> 
> I've edited the original post to make clear that this is a strawman proposal for discussion.
> 
> I also don't like spyware, but maybe an opt-in plugin that reports limited information could be acceptable?

Before talking about solutions, let's start at the beginning:
- what kind of information would you like to collect?
- what results would you expect from collecting and analysing these information?
- who should collect and analyse the data?

You probably will never get accurate numbers of users using Geany and what plugins they use:
- you will get data only from users participating
- such usage data might get outdated fast
- how do you count "use": users might start Geany once a month to edit a certain file, or start Geany in their spare time when coding, use Geany all day for professional work, used Geany some time in the past, users might have Geany installed but never tried it, ...

Even if you had such information, which problem would it really solve?

I agree that such information would be kind of interesting but, IMHO, nothing more than this.
Considering the amount of work it would require, for a technical implementation, the analysis of collected data but also to sort out all potential legal issues, I don't think it's worth.

Also to consider: maybe some, maybe many, maybe most users might refuse to use Geany if there was something like spyware added.
Even if it would be optional. 
Even if it would help the further development of Geany. 
Even if the collected data would be anonymous.

*I personally would uninstall if something like this would happen.*

> > I don't know what legal issues spyware may have, even if its possible to turn it off.
> 
> Many programs "check for updates". Software from larger companies do more, but of course, they likely have more monetary resources for legal battles than Geany does.

The UpdateChecker plugin (https://github.com/geany/geany-plugins/tree/master/updatechecker) already does checking for updates.
Actually, it already sends the Geany version to the server. But this information is neither stored nor analysed.
Even if, it wouldn't say much as this plugin is mainly only useful on Windows and maybe MacOS. The majority of users (assumption by me) doesn't need this plugin because they are using Linux and a package manager.


Alternative ideas to "spyware":

- Debian collects package usage statistics in the Popcon project. Geany: https://qa.debian.org/popcon.php?package=geany Geany-Plugins: https://qa.debian.org/popcon.php?package=geany-plugins
  Maybe this already helps? Note, these statistics represent only Debian users who opted in to the data collection.
  
- a few years ago, there was already the idea of creating a user survey to ask users how they use Geany. As far as I remember, it never happened because nobody has done it.

- maybe the issue tracker and mailing lists already give a good impression how and on what operating systems Geany is used. Obviously, the amount of users reporting issues and asking on mailing lists are only a fraction of the overall user base but might be somewhat representive.

**Comment by elextr** [comment]
@xiota 

> I've edited the original post to make clear that this is a strawman proposal for discussion.

Absolutely no problem to ask the question, having a discussion and a recorded decision is always a good thing.

> Many programs "check for updates".

Checking anonymously for updates without providing any information is fine, but its hard to build in usefully when every distro has its own source of upgrades.  

There is an updatechecker plugin, but it tells the user that Geany project has released a new version.  But it may not appear in many distros for (literally) years, so it just as likely annoys users than is actually useful. 

Its sending information about the users machine/usage that makes it spyware.

> People who ask for a feature to be "changed" would likely be satisfied with an option for their request.

Its surprising how many object to making their brilliant idea optional or additional, I have never really understood why.

@eht16 agree with everything you said.

Just a final rant if I may, statistics and machine "learning" and similar are all backward looking, they tell you what happened yesterday, not how to move forward to tomorrow.  And they are fragile, any change in the environment invalidates any information content there might have been in their data.  

Statistics on usage of a feature do not tell you how that usage will change if you change the UI or the exact functionality.  In fact they can potentially lead to dysfunctional behaviour "we don't need to improve function X because nobody uses it", but maybe they don't use it because it needs improvement.  Or maybe its simply not useful to most, but the stats don't tell you that.

 And statistics don't provide any guidance as to what is appropriate in the application or it ends up as a grab bag of disparate parts, to quote the maintainer of Scintilla "its as important to know what not to add as it is what to add" and no stats will help with that. 

End grumpy olde guy rant.
